draft-ietf-mls-protocol.md

```diff
@@ -914,6 +905,58 @@ np[0] -> 1       5[C]
 After performing these operations, the generator of the UpdatePath MUST
 delete the leaf_secret.
 
+## Adding and Removing Leaves
+
+In addition to the path-based updates to the tree described above, it is also
+necessary to add and remove leaves of the tree in order to reflect members being
+added to or removed from the group.  Leaves are always added and removed at the
+right edge of the tree: Either a new rightmost leaf is added, or the rightmost
+leaf is removed.  Nodes' parent/child node relationships are then updated to
+maintain the tree's left-balanced structure.  These operations are also known as
+_extending_ and _truncating_ the tree.
+
+To add a new leaf: Add leaf as the new rightmost leaf of the tree.  Add a blank
+parent node whose right child is the new leaf.  The parent is attached to the
+tree as the right child of the appropriate node to make the updated tree
+left-balanced (or set as a new root).  The former right child of the parent's
+parent becomes the parent's left child (or the old root becomes the left child).
+
+~~~~~
+                   _ <-- new parent              _
+                 __|_                          __|__
+                /    \                        /     \
+  X    ===>    X     |               ===>    X       _ <-- new parent
+ / \          / \    |                      / \     / \
+A   B        A   B   C <-- new leaf        A   B   C   D <-- new leaf
+~~~~~
+
+To remove the rightmost leaf: Remove the rightmost leaf node and its parent
+node.  If the parent node was the root of the tree, the parent node's left child
+is now the root of the tree.  Otherwise, set the right child of the parent's parent
+(formerly the parent itself) to be the parent's left child.
```

Comment:
I also suggest giving names to nodes. Also removing "(formerly the parent itself)" because I don't see what that means.
In the first example, the leaf is D, the parent is Z, the parent's parent is Y, so the sentence says that Y's right child is now Z's left child (so C). I don't get what "former parent" means here.
```suggestion
To remove the rightmost leaf: Remove the rightmost leaf node L and its parent
node P.  If P was the root of the tree, P's left child
is now the root of the tree.  Otherwise, set the right child of P's parent
to be the P's left child.
```